LSP: Add autocomplete & hover functionality, and jump to target name

MaartenStaa opened the pull request:
Fixes #52.

This PR adds the capability of sending auto-complete suggestions when in LSP mode. Currently it supports:

- Locally defined methods
- Imported symbols
- Local variables
- Function arguments
- Scoping said local variables and function arguments based on the position of the cursor (inside/outside a method)
- Automatically inserting/updating a `load` for exported symbols from another opened file

Not yet handled/open questions:

- The `kind` of previously loaded symbols is not always correct (i.e. it's always a `CompletionItemKind::METHOD` at the moment), because I'd need to resolve the file, which it doesn't do at the moment.
- Auto-complete for exported symbols when the cursor is in a `load` statement doesn't work yet, for the same reason of resolving the path.
- The path for generated `load` statements is based on Bazel, I don't know what this should look like for Buck and/or Starlark usage outside of these two build systems. This ties in with #51, which has been open for a while.
- The path for generated `load` statements doesn't handle external repository paths yet, this also ties in with #51.

I'm hoping with some feedback, and maybe a merge of the Bazel PR, we can get this to a mergeable state 🙂 looking forward to hearing your thoughts!

ndmitchell commented:
Sorry for the delay in looking at this - I was travelling.

How much of a problem is the use of kind = Method? What is the impact of that on a client such as VS Code? Just an incorrect icon in the completion list?

ndmitchell commented on starlark/src/lsp/server.rs:
```diff
@@ -587,6 +630,292 @@ impl<T: LspContext> Backend<T> {
+                        .read()
+                        .unwrap()
+                        .iter()
+                        .filter(|(&ref doc_uri, _)| doc_uri != &uri)
```
To confirm, this is the bit that encodes the relationship between files and load statements?

MaartenStaa replied:
Kind of. So this bit of code scans through exported symbols of all loaded files. The filter makes sure we _don't_ do that for the file we're generating completion items for, as the current file is handled separately through `Self::collect_symbols` above. It's mainly the `format_load_path` just below this that transforms a document URI to a path suitable for `load()`, in order to insert/amend a `load` statement upon accepting that completion item.

Note however that the mapping is one way, currently there is no way to take a path from a `load` statement and map it to a URI, e.g. to implement _go to definition_, or to check the type of the previously loaded symbol for that file.

MaartenStaa replied:
Resolving this since this is no longer true, `resolve_load` is a thing now.

ndmitchell commented on starlark/src/lsp/server.rs:
```diff
@@ -587,6 +630,292 @@ impl<T: LspContext> Backend<T> {
+                            .filter(|symbol| !symbols.contains_key(symbol.name))
+                            .map(|symbol| {
+                                // Construct the text edit to insert a load for this exported symbol.
+                                let text_edit = if let Some(existing_load) = loads.get(&load_path) {
```
Maybe the constructing text should be a separate helper function - getting a little unweildy.

MaartenStaa replied:
I imagine we want to do this in the `LspContext::render_as_load` we discussed below? If so, I guess that method would also need to accept a list of existing loaded items (including renames when applicable).

ndmitchell commented on starlark/src/lsp/server.rs:
```diff
@@ -28,6 +28,9 @@ use derivative::Derivative;
 use derive_more::Display;
 use dupe::Dupe;
 use dupe::OptionDupedExt;
+use gazebo::prelude::SliceExt;
```
We tend to import `prelude::*` rather than the specific instances

MaartenStaa replied:
Removed and no longer needed after refactoring.

ndmitchell commented on starlark/src/codemap.rs:
```diff
@@ -542,6 +542,13 @@ impl From<ResolvedSpan> for lsp_types::Range {
 }
 
 impl ResolvedSpan {
+    pub(crate) fn contains(&self, pos: LineCol) -> bool {
```
One line comment that this is entirely contained within

MaartenStaa commented:
Thanks for going through this @ndmitchell! I'll go through your comments some time this week and a) tidy up the code and b) split out the Bazel specific stuff to a separate PR. In that case the initial PR might have to omit completion of exported symbols from other files altogether, as completing those requires inserting the appropriate `load` statement.

Regarding the "method" type, yeah it's just an incorrect icon basically, or e.g. in Neovim it might have the type "method" as text instead of an icon. Not a huge deal obviously. 

Thanks for the tip about SaplingSCM, I'll take a look at that as well. For the follow-up PR, I might create a new branch based on both this one _and_ #51, thoughts?

ndmitchell commented:
I'm just about to go through #51 (the person who was working from this on our side left Meta, so it accidentally got lost, oops). Let me see.

ndmitchell commented:
For generating a `load` statement, I imagine you need a method on `LspContext` like `render_as_load(&self, target: &LspUrl, current_file: &LspUrl) -> anyhow::Result<String>` ? That seems quite reasonable to add in.

MaartenStaa commented:
That sounds like a good solution, yeah. Probably you want to be able to do the mapping in two directions:

- `LspContext::render_as_load(&self, target: &LspUrl, current_file: &LspUrl) -> anyhow::Result<String>`
- `LspContext::resolve_load(&self, target: &str, current_file: &LspUrl) -> anyhow::Result<LspUrl>`

MaartenStaa commented on starlark/src/lsp/server.rs:
```diff
@@ -332,6 +351,17 @@ impl<T: LspContext> Backend<T> {
+                    "[".to_string(),
+                ]),
+
+                // all_commit_characters: Some(vec![" ".to_string()]),
```
Removed, was a leftover from during my initial implementation.

MaartenStaa commented on starlark/src/lsp/server.rs:
```diff
@@ -587,6 +630,292 @@ impl<T: LspContext> Backend<T> {
+        codemap: &CodeMap,
+        ast: &AstStmt,
+        cursor_position: LineCol,
+        symbols: &mut HashMap<String, CompletionItem>,
```
Done, and extracted to its own module including tests.

ndmitchell commented:
Thanks! I'm going through the code now - plan to split pieces off and land them separately (those obvious bits) and slowly work until there is nothing left. Expect to see pieces of your code trickle out over the next day or so.

MaartenStaa commented:
> Thanks! I'm going through the code now - plan to split pieces off and land them separately (those obvious bits) and slowly work until there is nothing left. Expect to see pieces of your code trickle out over the next day or so.

@ndmitchell Awesome, looking forward to it. Feel free to ping me if any questions come up.

ndmitchell commented:
I've submitted all the changes apart from those to lsp/server.rs and a handful of lines to definition.rs. I'll update once they are all landed. For the changes to server:

* I think we need the LspContext so that it isn't Buck/Bazel specific, with the mapping from a URL to the load name and vice versa. Did anything happen to your LspContext suggestion above?
* As an extension, I note you aren't auto completing using names that are loaded by other open modules. E.g. if I have a separate file open, with `load("filename", "foo")` then offering to complete to `foo` makes sense, even though `foo` isn't exported by anything open.

Overall, looks good, and the automatic adding of load statements is super slick.

ndmitchell commented:
Other notes:

* We should probably have a method on LspContext to get a list of the global symbols.
* Completing keywords is quite nice - a hardcoded list of those should be fairly easy to add.

MaartenStaa commented:
@ndmitchell Nice! I merged in `origin/main` and resolved any conflicts.

> I think we need the LspContext so that it isn't Buck/Bazel specific, with the mapping from a URL to the load name and vice versa. Did anything happen to your LspContext suggestion above?

I don't think so, no. Not sure if the idea is to tackle that via #51 or not? I can try to come up with an initial implementation. One thought I had later on: should that actually be on the `LspContext`? Just wondering if it makes more sense to have Bazel as a "dialect", and let the LSP be aware of the Starlark dialect. I should probably add that I'm not sure what, if any, the actual differences are between Bazel and Buck and how they render load paths. If they're the same, then having it on the `LspContext` seems fine.

> As an extension, I note you aren't auto completing using names that are loaded by other open modules. E.g. if I have a separate file open, with load("filename", "foo") then offering to complete to foo makes sense, even though foo isn't exported by anything open.

That's a nice idea, I'll try to find some time to add that!

> We should probably have a method on LspContext to get a list of the global symbols.

Makes sense. I guess that would be the things under `starlark/src/stdlib`?

> Completing keywords is quite nice - a hardcoded list of those should be fairly easy to add.

Added!

> Overall, looks good, and the automatic adding of load statements is super slick.

Thanks! I was pretty happy with that too :)

MaartenStaa commented:
@ndmitchell I have now added the autocomplete options gathered from `load` statements in open files.

MaartenStaa commented:
@ndmitchell Added a few more things:

- I realised the autocomplete options from `load` statements in other files needed to have their path resolved and re-rendered correctly from the point-of-view of the target file
- So I implemented `render_as_load`
- Also added some fixes to `resolve_load`
- Added the workspace root as a parameter to both
- Added tests to both
- Also added autocompletion for global symbols (including docs!)

I'd appreciate it if you or someone else can take a critical look at `render_as_load` and `resolve_load`—I based my logic on how Bazel works, but don't know enough about Buck(2) to know whether it uses the same format of:
- In same folder: `:filename`
- Absolute: `//some/sub/folder:filename`
- In root: `//:filename`

(also, it doesn't really handle stuff in other workspaces, e.g. `@workspace//some:file`)

ndmitchell commented:
Thanks - hoping to take a look this weekend!

MaartenStaa commented:
@ndmitchell OK, so I went off the rails a bit this week 😅 added a bunch of stuff. You can see it from the commits above, but summarizing:

- Autocomplete takes context into account now, e.g. load paths/exported symbols in load statements, parameter names in function calls, target names in string literals
- Hover info, reusing much of the same infrastructure as was used for go to definition
- Go to definition (and thus hover info) as well now also works for target names, when those are defines with a macro/rules invocation in a build file with a `name` attribute
- Checks the used build system (support for Buck2 and Bazel) to gather information about known repositories and target names in packages. This also means `resolve_load` and such now also works for paths containing a `@repository//`-like prefix.

I appreciate that this probably makes reviewing this more complicated, so I'm sorry about that. Is there some (semi-)public channel (Slack, Discord) you have where maybe we can work together a bit more closely to be able to land this? e.g. I know for the Relay team they have a private channel in the GraphQL Discord server.

ndmitchell commented:
@MaartenStaa - might need a VC to ensure we move forward with this. Can you email me at ndmitchell -at- gmail.com and we can figure it out.

ndmitchell commented on starlark/src/syntax/grammar.lalrpop:
```diff
@@ -186,7 +186,7 @@ ExprStmt_: Stmt = <Test> => Stmt::Expression(<>);
 LoadStmt_: Stmt = LoadStmtInner => Stmt::Load(<>);
 
-LoadStmtInner: Load = "load" "(" <module:string> <args:("," <LoadStmtSyms>)+> ","? ")" => {
+LoadStmtInner: Load = "load" "(" <module:string> <args:("," <LoadStmtSyms>)*> ","? ")" => {
```
Why is this changing?

MaartenStaa replied:
Check this commit: https://github.com/facebookexperimental/starlark-rust/pull/68/commits/1d745b9b995b5438d5bad80fe33f2de3a518198e

In short, it's because the parser is too strict/not error correcting. If the input doesn't fully parse, then when an autocomplete is triggered, the parsed AST is the latest snapshot that did parse. As a result, if the input line is:

```
load("|")
```

Where `|` is the cursor, there's no way of knowing that the cursor is inside a `load` statement—as it's not syntactically correct.

ndmitchell commented on starlark/src/docs/mod.rs:
```diff
@@ -786,6 +800,30 @@ pub enum DocItem {
     Object(DocObject),
     Function(DocFunction),
     Property(DocProperty),
+    Param(DocParam),
```
This feels like the wrong place to put DocParam.

ndmitchell commented:
I've gone through this PR in consultation with @MaartenStaa:

* The build_system stuff and the string completion in load stuff should be in a separate binary, starlark-bazel. The suggestion is that is a temporary binary in starlark-rust repo while it is in active development, then fairly soon after, becomes a standalone project.
* The work to make the `load` statement less strict is valuable, but needs doing as a more general step to make the parser more tolerant for LSP. We can put that in later.
* I moved a lot of code that lived outside of starlark/lsp into starlark/lsp. The rough logic being that the code outside there should have strong, stable, thoughtful APIs we rely on. The stuff inside lsp is whatever LSP currently needs to get its job done.
* I tweaked some types in somewhat insignificant ways:
  * StringCompletion doesn't need to be passed the literal (if you care, capture it when giving back the function)
  * The environment completions might vary by URL, and DocModule is a better type for them.
  * The FixturesWith type didn't actually need the deferred CodeMap thing so I dropped it.

With that I have 6 diffs that are ready to land now (being reviewed and landed internally) then one diff on top which is the hover/completion stuff. Once the prerequisites land I'll publish that and share it back here (it is 95%+ the work of @MaartenStaa).

ndmitchell commented:
All the dependencies are now landed. https://github.com/facebookexperimental/starlark-rust/pull/79 is the code which is basically a version of this.

ndmitchell commented:
Given #79 is now landed, closing this diff as it is 90% landed. Please open follow up PR for the rest of the pieces. Thanks so much for your improvements!

stagnation commented:
Hooray, thank you! I checked with neovim and a simple bazel program to hopefully see if the `starlark-bazel` parts were included. The linting functionality that we've always had works just like before, and there are now more LSP functions implemented, go-to-definition can be called where it makes sense, loads and target references. But the landing location is typically just a (new) file with the token name, which I assume is expected for a bazel project. I'm happy to test-run `starlark-bazel` when it comes :)
